ignis/services/niri/service.py

```diff
@@ -108,65 +120,174 @@ def __listen_events(self) -> None:
+        event_data = list(json_data.values())[0]
+        match event_type:
+            case "KeyboardLayoutSwitched":
+                self.__update_current_layout(event_data)
+            case "KeyboardLayoutsChanged":
+                self.__update_keyboard_layouts(event_data)
+            case "WindowClosed":
+                self.__destroy_window(event_data)
+            case "WindowFocusChanged":
+                self.__update_window_focus(event_data)
+            case "WindowOpenedOrChanged":
+                self.__update_window(event_data)
+            case "WindowsChanged":
+                self.__update_windows(event_data)
+            case "WorkspaceActivated":
+                self.__update_active_workspace(event_data)
+            case "WorkspaceActiveWindowChanged":
+                self.__update_workspace_active_window(event_data)
+            case "WorkspacesChanged":
+                self.__update_workspaces(event_data)
+
+    def __update_current_layout(self, data: dict) -> None:
+        self._keyboard_layouts.sync({"current_idx": data["idx"]})
+        self.notify("keyboard_layouts")
+        self._keyboard_layouts.notify("current_name")
+
+    def __update_keyboard_layouts(self, data: dict) -> None:
+        self._keyboard_layouts.sync(data["keyboard_layouts"])
+        self.notify("keyboard_layouts")
+        self._keyboard_layouts.notify("current_name")
+
+    def __sort_windows(self) -> None:
+        self._windows = dict(sorted(self._windows.items()))
+
+    def __destroy_window(self, data: dict) -> None:
+        window = self._windows.pop(data["id"])
+        if window:
+            window.emit("destroyed")
+            self.__sort_windows()
+            self.notify("windows")
+
+    def __update_window(self, data: dict) -> None:
+        window_data = data["window"]
+        window = self._windows.get(window_data["id"], None)
+        if window is None:
+            window = NiriWindow(self)
+
+        window.sync(window_data)
+        self._windows[window_data["id"]] = window
+
+        if window.is_focused:
+            self._active_window.sync(window_data)
+
+        self.__sort_windows()
+
         self.notify("active-window")
+        self.notify("windows")
 
-    def __sync_active_output(self) -> None:
-        self._active_output = json.loads(self.send_command('"FocusedOutput"\n'))["Ok"][
-            "FocusedOutput"
-        ]
-        self.notify("active-output")
+    def __update_window_focus(self, data: dict) -> None:
+        # Id of the newly focused window, or None if no window is now focused.
+        id = data["id"]
+        for _id, window in self._windows.items():
+            # If a window became focused, it implies that all other windows
+            # are no longer focused. Go through existing windows and
+            # update their state accordingly.
+            window.sync({"is_focused": (window.id == id)})
+
+        if id:
+            self._active_window.sync(self._windows[id].data)
```

Comment:
```suggestion
        focused_id = data["id"]
        for window in self._windows.values():
            # If a window became focused, it implies that all other windows
            # are no longer focused. Go through existing windows and
            # update their state accordingly.
            window.sync({"is_focused": (window.id == focused_id)})

        if focused_id:
            self._active_window.sync(self._windows[focused_id].data)
```

Not critical, just code style improvements